sycl/include/sycl/ext/oneapi/experimental/device_architecture.hpp

```diff
@@ -24,92 +28,116 @@ enum class architecture {
   //   - the unique ID of the new architecture in SYCL RT source code to support
   //     querying the device architecture
   //
-  x86_64,
-  intel_cpu_spr,
-  intel_cpu_gnr,
-  intel_gpu_bdw,
-  intel_gpu_skl,
-  intel_gpu_kbl,
-  intel_gpu_cfl,
-  intel_gpu_apl,
+  x86_64 = 0x00000000,
+  //
+  // Intel CPU architectures
+  //
+  // The requirement for the unique ID for intel_cpu_* architectures below is:
+  // - the ID must start with 0x0 (to avoid the integer overflow)
+  // - then goes Intel's vendor ID from underlied backend (which is 8086)
+  // - the ID ends with the architecture ID from the DEVICE_IP_VERSION extension
+  //   of underlied backend
+  intel_cpu_spr = 0x08086008,
+  intel_cpu_gnr = 0x08086009,
+  //
+  // Intel GPU architectures
+  //
+  // The requirement for the unique ID for intel_gpu_* architectures below is:
+  // - the ID is GMDID of that architecture
+  intel_gpu_bdw = 0x02000000,
+  intel_gpu_skl = 0x02400009,
+  intel_gpu_kbl = 0x02404009,
+  intel_gpu_cfl = 0x02408009,
+  intel_gpu_apl = 0x0240c000,
   intel_gpu_bxt = intel_gpu_apl,
-  intel_gpu_glk,
-  intel_gpu_whl,
-  intel_gpu_aml,
-  intel_gpu_cml,
-  intel_gpu_icllp,
-  intel_gpu_ehl,
+  intel_gpu_glk = 0x02410000,
+  intel_gpu_whl = 0x02414000,
+  intel_gpu_aml = 0x02418000,
+  intel_gpu_cml = 0x0241c000,
+  intel_gpu_icllp = 0x02c00000,
+  intel_gpu_ehl = 0x02c08000,
   intel_gpu_jsl = intel_gpu_ehl,
-  intel_gpu_tgllp,
-  intel_gpu_rkl,
-  intel_gpu_adl_s,
+  intel_gpu_tgllp = 0x03000000,
+  intel_gpu_rkl = 0x03004000,
+  intel_gpu_adl_s = 0x03008000,
   intel_gpu_rpl_s = intel_gpu_adl_s,
-  intel_gpu_adl_p,
-  intel_gpu_adl_n,
-  intel_gpu_dg1,
-  intel_gpu_acm_g10,
+  intel_gpu_adl_p = 0x0300c000,
+  intel_gpu_adl_n = 0x03010000,
+  intel_gpu_dg1 = 0x03028000,
+  intel_gpu_acm_g10 = 0x030dc008,
   intel_gpu_dg2_g10 = intel_gpu_acm_g10,
-  intel_gpu_acm_g11,
+  intel_gpu_acm_g11 = 0x030e0005,
   intel_gpu_dg2_g11 = intel_gpu_acm_g11,
-  intel_gpu_acm_g12,
+  intel_gpu_acm_g12 = 0x030e4000,
   intel_gpu_dg2_g12 = intel_gpu_acm_g12,
-  intel_gpu_pvc,
-  intel_gpu_pvc_vg,
+  intel_gpu_pvc = 0x030f0007,
+  intel_gpu_pvc_vg = 0x030f4007,
+  //
   // NVIDIA architectures
-  nvidia_gpu_sm_50,
-  nvidia_gpu_sm_52,
-  nvidia_gpu_sm_53,
-  nvidia_gpu_sm_60,
-  nvidia_gpu_sm_61,
-  nvidia_gpu_sm_62,
-  nvidia_gpu_sm_70,
-  nvidia_gpu_sm_72,
-  nvidia_gpu_sm_75,
-  nvidia_gpu_sm_80,
-  nvidia_gpu_sm_86,
-  nvidia_gpu_sm_87,
-  nvidia_gpu_sm_89,
-  nvidia_gpu_sm_90,
+  //
+  // The requirement for the unique ID for nvidia_gpu_* architectures below is:
+  // - the ID must start with NVIDIA's vendor ID from underlied backend (which
+  //   is 0x10de)
```

Comment:
"vendor ID from underlied backend" is basically the PCI vendor ID, they are available here: https://devicehunt.com/all-pci-vendors. The reason I mentioned "underlined backend" is the you can query it from e.g. OpenCL. To make it more straight-forward, I removed backend mentioning, and replaced it with just "PIC vendor ID"

> Your strategy assumes that a vendor ID will never be the same as the top 4 hex digits of the GMDID of an Intel device. How confident are we that this will always be the case?

Good question! I introduced new requirements for "uniqueness" of architecture IDs in https://github.com/intel/llvm/pull/12752/commits/5a8efb0ca2b8a44dd4bace4d3163415949bda28f, can you please take a look? They all are in this comment: https://github.com/intel/llvm/pull/12752/files#diff-8cb9fcff5b559fb77322badec718a79d599471a6013ca84a6b34061daccafcb7R32-R47